internal/controller/state/dataplane/configuration.go

```diff
@@ -242,6 +319,137 @@ func buildStreamUpstreams(
 	return upstreams
 }
 
+// buildTCPUpstreams builds all TCP upstreams.
+func buildTCPUpstreams(
+	ctx context.Context,
+	gateway *graph.Gateway,
+	serviceResolver resolver.ServiceResolver,
+	ipFamily IPFamilyType,
+) []Upstream {
+	uniqueUpstreams := make(map[string]Upstream)
+
+	for _, l := range gateway.Listeners {
+		if !l.Valid || l.Source.Protocol != v1.TCPProtocolType {
+			continue
+		}
+
+		for _, route := range l.L4Routes {
+			if !route.Valid {
+				continue
+			}
+
+			br := route.Spec.BackendRef
+
+			if !br.Valid {
+				continue
+			}
+
+			gatewayNSName := client.ObjectKeyFromObject(gateway.Source)
+			if _, ok := br.InvalidForGateways[gatewayNSName]; ok {
+				continue
+			}
+
+			upstreamName := br.ServicePortReference()
+
+			if _, exist := uniqueUpstreams[upstreamName]; exist {
+				continue
+			}
+
+			var errMsg string
+
+			allowedAddressType := getAllowedAddressType(ipFamily)
+
+			eps, err := serviceResolver.Resolve(ctx, br.SvcNsName, br.ServicePort, allowedAddressType)
```

Comment:
This function call needs to be updated as it's missing the logger parameter - see the pattern in buildUpstreams:

`eps, err := svcResolver.Resolve(ctx, logger, br.SvcNsName, br.ServicePort, allowedAddressType)`